File: C-Sharp/FerryLegacy/JourneyBookingService.cs
using System.Collections.Generic;
using System.Linq;

namespace FerryLegacy
{
    public class JourneyBookingService
    {
        private TimeTables _timeTables;
        private Bookings _bookings;
        private readonly FerryAvailabilityService _ferryService;

        public JourneyBookingService(TimeTables timeTables, Bookings bookings, FerryAvailabilityService ferryService)
        {
            _timeTables = timeTables;
            _bookings = bookings;
            _ferryService = ferryService;
        }

        public bool CanBook(int journeyId, int passengers)
        {
            var timetables = _timeTables.All();
            var allEntries = timetables.SelectMany(x => x.Entries).OrderBy(x => x.Time).ToList();
            foreach (var timetable in allEntries)
            {
                var ferry = _ferryService.NextFerryAvailableFrom(timetable.OriginId, timetable.Time);

                if (timetable.Id == journeyId)
                {
                    var bookings = _bookings.All().Where(x => x.JourneyId == journeyId);
                    var seatsLeft = ferry.Passengers - bookings.Sum(x => x.Passengers);
                    return seatsLeft >= passengers;
                }
            }
            return false;
        }

        public void Book(Booking booking)
        {
            _bookings.Add(booking);
        }

        public IEnumerable<Booking> GetAllBookings()
        {
            return _bookings.All();
        }
    }
}

File: java/src/main/java/ferry/booking/JourneyBookingService.java
package ferry.booking;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class JourneyBookingService {

    private TimeTables timeTables;
    private Bookings bookings;
    private final FerryAvailabilityService ferryService;

    public JourneyBookingService(TimeTables timeTables, Bookings bookings, FerryAvailabilityService ferryService) {
        this.timeTables = timeTables;
        this.bookings = bookings;
        this.ferryService = ferryService;
    }

    public boolean canBook(int journeyId, int passengers) {
        List<TimeTable> timetables = timeTables.all();
        List<TimeTableEntry> allEntries = new ArrayList<TimeTableEntry>();
        for (TimeTable tt : timetables) {
            allEntries.addAll(tt.entries);
        }
        Collections.sort(allEntries, new Comparator<TimeTableEntry>() {

            @Override
            public int compare(TimeTableEntry tte1, TimeTableEntry tte2) {
                return Long.compare(tte1.time, tte2.time);
            }
        });

        for (TimeTableEntry timetable : allEntries) {
            Ferry ferry = ferryService.nextFerryAvailableFrom(timetable.originId, timetable.time);

            if (timetable.id == journeyId) {
                List<Booking> journeyBookings = new ArrayList<>();
                for (Booking x : bookings.all()) {
                    if (x.journeyId == journeyId) {
                        journeyBookings.add(x);
                    }
                }
                int pax = 0;
                for (Booking x : bookings.all()) {
                    pax += x.passengers;
                }
                int seatsLeft = ferry.passengers - pax;
                return seatsLeft >= passengers;
            }
        }
        return false;
    }

    public void book(Booking booking) {
        bookings.add(booking);
    }

    public List<Booking> getAllBookings() {
        return bookings.all();
    }
}
